src/main/java/io/github/syst3ms/skriptparser/registration/SkriptRegistration.java

```diff
@@ -527,11 +567,47 @@ public void register() {
             registerer.addHandledEvent(this.c);
         }
 
+        /**
+         * Set the context this event can handle
+         * @param contexts the contexts
+         * @return the registrar
+         */
         @SafeVarargs
         public final EventRegistrar<T> setHandledContexts(Class<? extends TriggerContext>... contexts) {
             this.handledContexts = contexts;
             return this;
         }
+
+        /**
+         * Registers a {@link ContextValue}
+         * @param context the context this value appears in
+         * @param type the returned type of this context value
+         * @param name the name of this context value (used in the suffix)
+         * @param contextFunction the function that needs to be applied in order to get the context value
+         * @param <C> the context class
+         * @param <T2> the type class
+         * @return the registrar
+         */
+        public final <C extends TriggerContext, T2> EventRegistrar<T> addContextValue(Class<C> context, Class<T2> type, String name, Function<C, T2[]> contextFunction) {
+            contextValues.add(new ContextValue<>(context, type, name, (Function<TriggerContext, T2[]>) contextFunction));
+            return this;
+        }
+
+        /**
+         * Registers a {@link ContextValue}
+         * @param context the context this value appears in
+         * @param type the returned type of this context value
+         * @param name the name of this context value (used in the suffix)
+         * @param contextFunction the function that needs to be applied in order to get the context value
+         * @param time whether this happens in the present, past or future
+         * @param <C> the context class
+         * @param <T2> the type class
+         * @return the registrar
+         */
+        public final <C extends TriggerContext, T2> EventRegistrar<T> addContextValue(Class<C> context, Class<T2> type, String name, Function<C, T2[]> contextFunction, ContextValueTime time) {
```

Comment:
The generic type should be T everywhere, this code probably errors as it is.

Reply:
It doesn't error. I did it this way because it the EventRegistrar itself uses T as a generic type. I couldn't overwrite it in the method signature since it needs to return an EventRegistrar with that specific generic type inside it.

Reply:
But the generic type T isn't declared anywhere in your method though.

Reply:
It is in the class? Or does method generic types override ALL existing generic types of that class even if they aren't specified in the method's generic types?

Reply:
Oh gee, you're right, sorry.